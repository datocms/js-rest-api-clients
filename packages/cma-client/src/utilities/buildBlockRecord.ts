import * as Utils from '@datocms/rest-client-utils';
import type { NewBlockInRequest } from '../fieldTypes';
import type * as ApiTypes from '../generated/ApiTypes';
import { Item } from '../generated/resources';
import type { ItemTypeDefinition } from './itemDefinition';

type NoInfer<T> = [T][T extends any ? 0 : never];

export function buildBlockRecord<
  D extends ItemTypeDefinition = ItemTypeDefinition,
>(body: ApiTypes.ItemUpdateSchema<NoInfer<D>>): NewBlockInRequest<NoInfer<D>> {
  return Utils.serializeRequestBody<{
    data: NewBlockInRequest<NoInfer<D>>;
  }>(body, {
    type: Item.TYPE,
    attributes: '*',
    relationships: ['item_type'],
  }).data;
}
