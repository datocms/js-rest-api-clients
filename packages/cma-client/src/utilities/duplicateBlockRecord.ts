import type {
  ItemWithOptionalIdAndMeta,
  NewBlockInRequest,
} from '../fieldTypes';
import type { ItemTypeDefinition } from './itemDefinition';
import { mapBlocksInNonLocalizedFieldValue } from './recursiveBlocks';
import type { SchemaRepository } from './schemaRepository';

type NoInfer<T> = [T][T extends any ? 0 : never];

export async function duplicateBlockRecord<
  D extends ItemTypeDefinition = ItemTypeDefinition,
>(
  existingBlock: ItemWithOptionalIdAndMeta<NoInfer<D>>,
  schemaRepository: SchemaRepository,
): Promise<NewBlockInRequest<NoInfer<D>>> {
  const { __itemTypeId, type, attributes, relationships } = existingBlock;

  const itemType = await schemaRepository.getRawItemTypeById(
    existingBlock.relationships.item_type.data.id,
  );

  const newBlock = {
    __itemTypeId,
    type,
    relationships,
    attributes,
  } as NewBlockInRequest;

  const fields = await schemaRepository.getRawItemTypeFields(itemType);

  for (const field of fields) {
    newBlock.attributes[field.attributes.api_key] =
      await mapBlocksInNonLocalizedFieldValue(
        newBlock.attributes[field.attributes.api_key],
        field.attributes.field_type,
        schemaRepository,
        (block, path) => {
          if (typeof block === 'string') {
            throw new Error(
              `Block cannot be duplicated as it contains nested block at ${path.join('.')} that is expressed as ID (${block}) instead of full object!`,
            );
          }

          const { id, meta, ...blockWithoutIdAndMeta } = block;

          return blockWithoutIdAndMeta as NewBlockInRequest;
        },
      );
  }

  return newBlock as NewBlockInRequest<NoInfer<D>>;
}
