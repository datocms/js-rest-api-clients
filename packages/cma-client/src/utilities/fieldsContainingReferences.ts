import type * as SchemaTypes from '../generated/SchemaTypes';
import type * as SimpleSchemaTypes from '../generated/SimpleSchemaTypes';

export function modelIdsReferencedInField(
  field: SchemaTypes.Field | SimpleSchemaTypes.Field,
) {
  const attributes = 'attributes' in field ? field.attributes : field;

  switch (attributes.field_type) {
    case 'link': {
      return attributes.validators.item_item_type.item_types;
    }
    case 'links': {
      return attributes.validators.items_item_type.item_types;
    }
    case 'structured_text': {
      return attributes.validators.structured_text_links.item_types;
    }
    default: {
      return [];
    }
  }
}

export function blockModelIdsReferencedInField(
  field: SchemaTypes.Field | SimpleSchemaTypes.Field,
) {
  const attributes = 'attributes' in field ? field.attributes : field;

  switch (attributes.field_type) {
    case 'single_block': {
      return attributes.validators.single_block_blocks.item_types;
    }
    case 'rich_text': {
      return attributes.validators.rich_text_blocks.item_types;
    }
    case 'structured_text': {
      return [
        ...attributes.validators.structured_text_blocks.item_types,
        ...(attributes.validators.structured_text_inline_blocks
          ? attributes.validators.structured_text_inline_blocks.item_types
          : []),
      ];
    }
    default: {
      return [];
    }
  }
}
