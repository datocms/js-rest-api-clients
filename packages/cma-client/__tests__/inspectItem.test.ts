import {
  type ApiTypes,
  type ItemTypeDefinition,
  type RawApiTypes,
  type ToItemDefinitionAsRequest,
  buildBlockRecord,
} from '../src';
import { inspectItem } from '../src/utilities/inspectItem';

type EnvironmentSettings = { locales: 'en' | 'it' | 'fr' };

type ComprehensiveModel = ItemTypeDefinition<
  EnvironmentSettings,
  'O9BXqTayQ_Wf-Yw6d863LQ',
  {
    string_field: { type: 'string' };
    text_field: { type: 'text' };
    boolean_true: { type: 'boolean' };
    boolean_false: { type: 'boolean' };
    boolean_null: { type: 'boolean' };
    integer_field: { type: 'integer' };
    float_field: { type: 'float' };
    color_field: { type: 'color' };
    date_field: { type: 'date' };
    datetime_field: { type: 'date_time' };
    file_field: { type: 'file' };
    gallery_field: { type: 'gallery' };
    json_field: { type: 'json' };
    lat_lon_field: { type: 'lat_lon' };
    link_field: { type: 'link' };
    links_field: { type: 'links' };
    slug_field: { type: 'slug' };
    seo_field: { type: 'seo' };
    video_field: { type: 'video' };
    structured_text: {
      type: 'structured_text';
      blocks: Block;
      inline_blocks: Block;
    };
    single_block: {
      type: 'single_block';
      blocks: Block;
    };
    rich_text: {
      type: 'rich_text';
      blocks: Block;
    };
    null_string: { type: 'string' };
    null_file: { type: 'file' };
    empty_gallery: { type: 'gallery' };
    null_link: { type: 'link' };
    empty_links: { type: 'links' };
  }
>;

type LocalizedModel = ItemTypeDefinition<
  EnvironmentSettings,
  'O9BXqTayQ_Wf-Yw6d863LQ',
  {
    string_field: { type: 'string'; localized: true };
    text_field: { type: 'text'; localized: true };
    integer_field: { type: 'integer'; localized: true };
    file_field: { type: 'file'; localized: true };
    structured_text: {
      type: 'structured_text';
      blocks: Block;
      inline_blocks: Block;
      localized: true;
    };
    seo_field: { type: 'seo'; localized: true };
    boolean_true: { type: 'boolean' };
  }
>;

type Block = ItemTypeDefinition<
  EnvironmentSettings,
  'NOYldjT6TtKXXjZ8efU-dw',
  {
    title: { type: 'string' };
    structured_text: {
      type: 'structured_text';
      blocks: Block;
      inline_blocks: Block;
    };
  }
>;

describe('inspectItem', () => {
  it('inspects an item with all field types', () => {
    const item: any = {
      type: 'item',
      id: 'IdMLV2GJTXyQ0Bfns7R4IQ',
      relationships: {
        item_type: {
          data: {
            type: 'item_type',
            id: 'O9BXqTayQ_Wf-Yw6d863LQ',
          },
        },
      },
      meta: {
        created_at: '2023-01-01T00:00:00Z',
        updated_at: '2023-01-01T00:00:00Z',
        published_at: null,
        first_published_at: null,
        publication_scheduled_at: null,
        unpublishing_scheduled_at: null,
        is_valid: true,
        is_current_version_valid: true,
        is_published_version_valid: false,
        stage: 'draft',
        current_version: '1',
        status: 'draft',
        has_children: false,
      },
      attributes: {
        // String field
        string_field: 'Sample string',

        // Text field
        text_field:
          'This is a longer text content that might span multiple lines',

        // Boolean fields
        boolean_true: true,
        boolean_false: false,
        boolean_null: null,

        // Number fields
        integer_field: 42,
        float_field: Math.PI,

        // Color field
        color_field: {
          red: 255,
          green: 128,
          blue: 0,
          alpha: 255,
        },

        // Date and DateTime fields
        date_field: '2023-12-25',
        datetime_field: '2023-12-25T10:30:00Z',

        // File field
        file_field: {
          upload_id: 'P4qKpnFfTR6Tc96-NwceLQ',
          alt: 'Sample image alt text',
          title: 'Sample image title',
          custom_data: {
            photographer: 'John Doe',
            location: 'Paris',
          },
          focal_point: {
            x: 0.5,
            y: 0.3,
          },
        },

        // Gallery field
        gallery_field: [
          {
            upload_id: 'RvLIJ-z0T32S2hd71zBxnQ',
            alt: 'First gallery image',
            title: 'Gallery Image 1',
            custom_data: {},
            focal_point: null,
          },
          {
            upload_id: 'a0QsFvzMQKCH1h-axboXKQ',
            alt: null,
            title: 'Gallery Image 2',
            custom_data: {
              category: 'landscape',
            },
            focal_point: {
              x: 0.7,
              y: 0.4,
            },
          },
        ],

        // JSON field
        json_field: JSON.stringify({
          nested: {
            data: 'value',
            number: 123,
            array: [1, 2, 3],
          },
        }),

        // Lat/Lon field
        lat_lon_field: {
          latitude: 48.8566,
          longitude: 2.3522,
        },

        // Link field
        link_field: 'FicV5CxCSQ6yOrgfwRoiKA',

        // Links field
        links_field: ['NVPfHLHQQZe_rDi4_e6M2w', 'VY3VzQFUS9C1ITDkG48Sgw'],

        // Slug field
        slug_field: 'sample-slug-value',

        // SEO field
        seo_field: {
          title: 'SEO Title',
          description: 'SEO Description for better search engine optimization',
          image: null,
          no_index: false,
          twitter_card: null,
        },

        // Video field
        video_field: {
          provider: 'youtube',
          provider_uid: 'dQw4w9WgXcQ',
          url: 'https://www.youtube.com/watch?v=dQw4w9WgXcQ',
          title: 'Sample Video Title',
          thumbnail_url:
            'https://img.youtube.com/vi/dQw4w9WgXcQ/maxresdefault.jpg',
          width: 1920,
          height: 1080,
        },

        structured_text: {
          schema: 'dast',
          document: {
            type: 'root',
            children: [
              {
                type: 'heading',
                level: 1,
                children: [{ type: 'span', value: 'Main Title' }],
              },
              {
                type: 'paragraph',
                children: [
                  { type: 'span', value: 'This is a ' },
                  { type: 'span', marks: ['strong'], value: 'bold' },
                  { type: 'span', value: ' paragraph.' },
                ],
              },
              {
                type: 'list',
                style: 'bulleted',
                children: [
                  {
                    type: 'listItem',
                    children: [
                      {
                        type: 'paragraph',
                        children: [{ type: 'span', value: 'First item' }],
                      },
                    ],
                  },
                  {
                    type: 'listItem',
                    children: [
                      {
                        type: 'paragraph',
                        children: [{ type: 'span', value: 'Second item' }],
                      },
                    ],
                  },
                ],
              },
              {
                type: 'block',
                item: 'block-123',
              },
              {
                type: 'block',
                item: buildBlockRecord<Block>({
                  item_type: {
                    type: 'item_type',
                    id: 'NOYldjT6TtKXXjZ8efU-dw',
                  },
                  title: 'Foobar',
                }),
              },
              {
                type: 'code',
                language: 'javascript',
                code: 'console.log("hello");',
              },
            ],
          },
        },

        single_block: buildBlockRecord<Block>({
          item_type: {
            type: 'item_type',
            id: 'NOYldjT6TtKXXjZ8efU-dw',
          },
          title: 'Foobar',
        }),

        rich_text: [
          'RjJIdPrFSwuxSBQQySDjxw',
          buildBlockRecord<Block>({
            item_type: {
              type: 'item_type',
              id: 'NOYldjT6TtKXXjZ8efU-dw',
            },
            title: 'Foobar',
          }),
        ],

        // Null values
        null_string: null,
        null_file: null,
        empty_gallery: [],
        null_link: null,
        empty_links: [],
      },
    };

    const output = inspectItem(item, 80);

    expect(output).toMatchSnapshot();
  });

  it('RawApiTypes.Item', () => {
    const item: ApiTypes.ItemCreateSchema<
      ToItemDefinitionAsRequest<ComprehensiveModel>
    > = {
      __itemTypeId: 'O9BXqTayQ_Wf-Yw6d863LQ',
      single_block: buildBlockRecord<Block>({
        item_type: {
          type: 'item_type',
          id: 'NOYldjT6TtKXXjZ8efU-dw',
        },
        title: 'Foobar',
      }),
      lat_lon_field: {
        latitude: 48.8566,
        longitude: 2.3522,
      },
      item_type: { type: 'item_type', id: 'O9BXqTayQ_Wf-Yw6d863LQ' },
    };

    const output = inspectItem(item, 80);

    expect(output).toMatchSnapshot();
  });

  it('inspects localized fields', () => {
    const item: RawApiTypes.Item<ToItemDefinitionAsRequest<LocalizedModel>> = {
      type: 'item',
      id: 'localizedTestId',
      relationships: {
        item_type: {
          data: {
            type: 'item_type',
            id: 'O9BXqTayQ_Wf-Yw6d863LQ',
          },
        },
      },
      meta: {
        created_at: '2023-01-01T00:00:00Z',
        updated_at: '2023-01-01T00:00:00Z',
        published_at: null,
        first_published_at: null,
        publication_scheduled_at: null,
        unpublishing_scheduled_at: null,
        is_valid: true,
        is_current_version_valid: true,
        is_published_version_valid: false,
        stage: 'draft',
        current_version: '1',
        status: 'draft',
        has_children: false,
      },
      attributes: {
        // Localized string field
        string_field: {
          en: 'Hello World',
          it: 'Ciao Mondo',
          fr: 'Bonjour le Monde',
        },

        // Localized text field
        text_field: {
          en: 'This is a long text in English',
          it: 'Questo è un testo lungo in italiano',
        },

        // Localized integer field
        integer_field: {
          en: 42,
          it: 24,
        },

        // Localized file field
        file_field: {
          en: {
            upload_id: 'english-file-id',
            alt: 'English alt text',
            title: 'English file title',
            custom_data: {},
            focal_point: null,
          },
          it: {
            upload_id: 'italian-file-id',
            alt: 'Testo alternativo italiano',
            title: 'Titolo file italiano',
            custom_data: { locale: 'it' },
            focal_point: { x: 0.6, y: 0.4 },
          },
        },

        // Localized structured text field
        structured_text: {
          en: {
            schema: 'dast',
            document: {
              type: 'root',
              children: [
                {
                  type: 'paragraph',
                  children: [{ type: 'span', value: 'English content' }],
                },
              ],
            },
          },
          it: {
            schema: 'dast',
            document: {
              type: 'root',
              children: [
                {
                  type: 'paragraph',
                  children: [{ type: 'span', value: 'Contenuto italiano' }],
                },
              ],
            },
          },
        },

        // Non-localized field for comparison
        boolean_true: true,

        // Localized SEO field
        seo_field: {
          en: {
            title: 'English SEO Title',
            description: 'English SEO description',
            image: null,
            no_index: false,
            twitter_card: null,
          },
          it: {
            title: 'Titolo SEO Italiano',
            description: 'Descrizione SEO italiana',
            image: null,
            no_index: false,
            twitter_card: null,
          },
        },
      },
    };

    const output = inspectItem(item, 80);

    expect(output).toMatchSnapshot();
  });
});
