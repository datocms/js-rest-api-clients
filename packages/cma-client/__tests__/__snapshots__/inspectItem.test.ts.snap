// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inspectItem ApiTypes.ItemCreateSchema 1`] = `
"└ Item (item_type: "O9BXqTayQ_Wf-Yw6d863LQ")
  ├ single_block
  │ └ Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │   └ title: "Foobar"
  └ lat_lon_field
    ├ latitude: 48.8566
    └ longitude: 2.3522
"
`;

exports[`inspectItem ApiTypes.ItemCreateSchema 2`] = `
"└ Item "IdMLV2GJTXyQ0Bfns7R4IQ" (item_type: "O9BXqTayQ_Wf-Yw6d863LQ")
  ├ string_field: "Sample string"
  ├ text_field: "This is a longer text content that might span multiple lines"
  ├ boolean_true: true
  ├ boolean_false: false
  ├ boolean_null: null
  ├ integer_field: 42
  ├ float_field: 3.141592653589793
  ├ color_field: #FF8000
  ├ date_field: 2023-12-25
  ├ datetime_field: 2023-12-25T10:30:00Z
  ├ file_field
  │ ├ upload_id: "P4qKpnFfTR6Tc96-NwceLQ"
  │ ├ alt: "Sample image alt text"
  │ ├ title: "Sample image title"
  │ ├ custom_data: {"photographer":"John Doe","location":"Paris"}
  │ └ focal_point: x=50% y=30%
  ├ gallery_field
  │ ├ [0]
  │ │ ├ upload_id: "RvLIJ-z0T32S2hd71zBxnQ"
  │ │ ├ alt: "First gallery image"
  │ │ └ title: "Gallery Image 1"
  │ └ [1]
  │   ├ upload_id: "a0QsFvzMQKCH1h-axboXKQ"
  │   ├ title: "Gallery Image 2"
  │   ├ custom_data: {"category":"landscape"}
  │   └ focal_point: x=70% y=40%
  ├ json_field: {"nested":{"data":"value","number":123,"array":[1,2,3]}}
  ├ lat_lon_field
  │ ├ latitude: 48.8566
  │ └ longitude: 2.3522
  ├ link_field: "FicV5CxCSQ6yOrgfwRoiKA"
  ├ links_field
  │ ├ [0]: "NVPfHLHQQZe_rDi4_e6M2w"
  │ └ [1]: "VY3VzQFUS9C1ITDkG48Sgw"
  ├ slug_field: "sample-slug-value"
  ├ seo_field
  │ ├ title: "SEO Title"
  │ ├ description: "SEO Description for better search engine optimization"
  │ └ no_index: false
  ├ video_field
  │ └ title: "Sample Video Title"
  ├ structured_text
  │ ├ heading (level: 1)
  │ │ └ span "Main Title"
  │ ├ paragraph
  │ │ ├ span "This is a "
  │ │ ├ span (marks: strong) "bold"
  │ │ └ span " paragraph."
  │ ├ list (style: bulleted)
  │ │ ├ listItem
  │ │ │ └ paragraph
  │ │ │   └ span "First item"
  │ │ └ listItem
  │ │   └ paragraph
  │ │     └ span "Second item"
  │ ├ block "block-123"
  │ ├ block
  │ │ └ Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │ │   └ title: "Foobar"
  │ └ code (language: "javascript") "console.log("hello");"
  ├ single_block
  │ └ Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │   └ title: "Foobar"
  ├ rich_text
  │ ├ [0] "RjJIdPrFSwuxSBQQySDjxw"
  │ └ [1] Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │   └ title: "Foobar"
  ├ null_string: null
  ├ null_file: null
  ├ empty_gallery: []
  ├ null_link: null
  └ empty_links: []
"
`;

exports[`inspectItem inspects an item with all field types 1`] = `
"└ Item "IdMLV2GJTXyQ0Bfns7R4IQ" (item_type: "O9BXqTayQ_Wf-Yw6d863LQ")
  ├ string_field: "Sample string"
  ├ text_field: "This is a longer text content that might span multiple lines"
  ├ boolean_true: true
  ├ boolean_false: false
  ├ boolean_null: null
  ├ integer_field: 42
  ├ float_field: 3.141592653589793
  ├ color_field: #FF8000
  ├ date_field: 2023-12-25
  ├ datetime_field: 2023-12-25T10:30:00Z
  ├ file_field
  │ ├ upload_id: "P4qKpnFfTR6Tc96-NwceLQ"
  │ ├ alt: "Sample image alt text"
  │ ├ title: "Sample image title"
  │ ├ custom_data: {"photographer":"John Doe","location":"Paris"}
  │ └ focal_point: x=50% y=30%
  ├ gallery_field
  │ ├ [0]
  │ │ ├ upload_id: "RvLIJ-z0T32S2hd71zBxnQ"
  │ │ ├ alt: "First gallery image"
  │ │ └ title: "Gallery Image 1"
  │ └ [1]
  │   ├ upload_id: "a0QsFvzMQKCH1h-axboXKQ"
  │   ├ title: "Gallery Image 2"
  │   ├ custom_data: {"category":"landscape"}
  │   └ focal_point: x=70% y=40%
  ├ json_field: {"nested":{"data":"value","number":123,"array":[1,2,3]}}
  ├ lat_lon_field
  │ ├ latitude: 48.8566
  │ └ longitude: 2.3522
  ├ link_field: "FicV5CxCSQ6yOrgfwRoiKA"
  ├ links_field
  │ ├ [0]: "NVPfHLHQQZe_rDi4_e6M2w"
  │ └ [1]: "VY3VzQFUS9C1ITDkG48Sgw"
  ├ slug_field: "sample-slug-value"
  ├ seo_field
  │ ├ title: "SEO Title"
  │ ├ description: "SEO Description for better search engine optimization"
  │ └ no_index: false
  ├ video_field
  │ └ title: "Sample Video Title"
  ├ structured_text
  │ ├ heading (level: 1)
  │ │ └ span "Main Title"
  │ ├ paragraph
  │ │ ├ span "This is a "
  │ │ ├ span (marks: strong) "bold"
  │ │ └ span " paragraph."
  │ ├ list (style: bulleted)
  │ │ ├ listItem
  │ │ │ └ paragraph
  │ │ │   └ span "First item"
  │ │ └ listItem
  │ │   └ paragraph
  │ │     └ span "Second item"
  │ ├ block "block-123"
  │ ├ block
  │ │ └ Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │ │   └ title: "Foobar"
  │ └ code (language: "javascript") "console.log("hello");"
  ├ single_block
  │ └ Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │   └ title: "Foobar"
  ├ rich_text
  │ ├ [0] "RjJIdPrFSwuxSBQQySDjxw"
  │ └ [1] Item (item_type: "NOYldjT6TtKXXjZ8efU-dw")
  │   └ title: "Foobar"
  ├ null_string: null
  ├ null_file: null
  ├ empty_gallery: []
  ├ null_link: null
  └ empty_links: []
"
`;

exports[`inspectItem inspects localized fields 1`] = `
"└ Item "localizedTestId" (item_type: "O9BXqTayQ_Wf-Yw6d863LQ")
  ├ string_field
  │ ├ en: "Hello World"
  │ ├ fr: "Bonjour le Monde"
  │ └ it: "Ciao Mondo"
  ├ text_field
  │ ├ en: "This is a long text in English"
  │ └ it: "Questo è un testo lungo in italiano"
  ├ integer_field
  │ ├ en: 42
  │ └ it: 24
  ├ file_field
  │ ├ en
  │ │ ├ upload_id: "english-file-id"
  │ │ ├ alt: "English alt text"
  │ │ └ title: "English file title"
  │ └ it
  │   ├ upload_id: "italian-file-id"
  │   ├ alt: "Testo alternativo italiano"
  │   ├ title: "Titolo file italiano"
  │   ├ custom_data: {"locale":"it"}
  │   └ focal_point: x=60% y=40%
  ├ structured_text
  │ ├ en
  │ │ └ paragraph
  │ │   └ span "English content"
  │ └ it
  │   └ paragraph
  │     └ span "Contenuto italiano"
  ├ boolean_true: true
  └ seo_field
    ├ en
    │ ├ title: "English SEO Title"
    │ ├ description: "English SEO description"
    │ └ no_index: false
    └ it
      ├ title: "Titolo SEO Italiano"
      ├ description: "Descrizione SEO italiana"
      └ no_index: false
"
`;
